sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs

```diff
@@ -0,0 +1,175 @@
+
+        /*To be used for General Login Scheme*/
+        public TokenCredentials(string username, string password)
+        {
+            _authHeader = EncodeTo64(username + ":" + password);
+        }
+
+        /*To be used for exchanging AAD Tokens for ACR Tokens*/
+        public TokenCredentials()
+        {
+            _authHeader = null;
+        }
+
+        public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request");
+            }
+            if (_authHeader != null)
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", _authHeader);
+            }
+            await base.ProcessHttpRequestAsync(request, cancellationToken);
+        }
+
+        static public string EncodeTo64(string toEncode)
+        {
+            byte[] toEncodeAsBytes = System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode);
+            string returnValue = System.Convert.ToBase64String(toEncodeAsBytes);
+            return returnValue;
+        }
+    }
+
+    /// <summary>
+    /// This allows us to have refresh token chains. For example if an access token needs a refresh
+    /// it can use the refresh token which can be refreshed using an aad access token which can be
+    /// refreshed using an aad refresh token which can be obtained using service principals.This
+    /// will all be done internally and thus abstracts much of the checking away.
+    /// </summary>
+    public class AuthToken
+    {
+
+        public delegate string acquireCallback();
+        private static readonly JwtSecurityTokenHandler JwtSecurityClient = new JwtSecurityTokenHandler();
+
+        // Constant to refresh tokens slightly before they are to expire guarding against possible latency related crashes
+        private TimeSpan LATENCY_SAFETY { get; set; } = TimeSpan.FromMinutes(2);
+
+        public string Value { get; set; }
+        public DateTime Expiration { get; set; }
+        public acquireCallback RefreshFn { get; set; }
+
+        public AuthToken(string token)
+        {
+            Value = token;
+            Expiration = JwtSecurityClient.ReadToken(Value).ValidTo;
+        }
+
+        public AuthToken(string token, acquireCallback refreshFn) : this(token)
+        {
+            RefreshFn = refreshFn;
+        }
+
+        //Extensibility purposes
+        protected AuthToken() { }
+
+
+        /* Returns true if refresh was successful. */
+        public bool Refresh()
+        {
+            if (RefreshFn == null)
+            {
+                return false;
+            }
+            Value = RefreshFn();
+            Expiration = JwtSecurityClient.ReadToken(Value).ValidTo;
+
+            return true;
+        }
+
+        public bool NeedsRefresh()
+        {
+            return Expiration < DateTime.UtcNow.Add(LATENCY_SAFETY);
```

Comment:
Question: Are both dates represented using the same date format?

Reply:
Yes, they should be. They are both DateTime objects and they should both be utc based. I can try to verify whether the jwt token's time is utc based.